preprocess/foldfilter_main.cc

```diff
@@ -0,0 +1,320 @@
+		for (int32_t pos_next = pos_cut_end; pos_cut_end < length; pos_cut_end = pos_next) {
+			// When we're not skipping delimiters, don't send more bytes than
+			// column_width in total a single line, even though we try to keep
+			// the delimiters together.
+			if (options.keep_delimiters && pos_cut_end - pos_last_cut >= static_cast<int32_t>(options.column_width))
+				break;
+
+			U8_NEXT(line.data(), pos_next, length, character);
+
+			if (character < 0)
+				throw utf8::NotUTF8Exception(line);
+
+			// First character after pos_cut is probably a delimiter, unless
+			// we did a hard stop in the middle of a word, and we're not keeping
+			// the delimiters.
+			if (find_delimiter(options.delimiters, character) == not_found)
+				break;
+		}
+
+		if (options.keep_delimiters) {
+			out_lines.push_back(line.substr(pos_last_cut, pos_cut_end - pos_last_cut));
+			out_delimiters.emplace_back("");
+		} else {
+			out_lines.push_back(line.substr(pos_last_cut, pos_cut - pos_last_cut));
+			out_delimiters.emplace_back(line.substr(pos_cut, pos_cut_end - pos_cut).data(), pos_cut_end - pos_cut);
+		}
+
+		pos_last_cut = pos_cut_end;
+		pos = pos_cut_end;
+	}
+
+	// Push out any trailing bits. Or the empty bit.
+	if (pos_last_cut < pos || pos == 0) {
+		out_lines.push_back(line.substr(pos_last_cut, pos - pos_last_cut));
+		out_delimiters.push_back("");
+	}
+
+	return make_pair(out_lines, out_delimiters);
+}
+
+int usage(char **argv) {
+	cerr << "usage: " << argv[0] << " [-w width] [-s] [-h] command [command-args ...]\n"
+		    "\n"
+		    "Options:\n"
+		    "  -h        Display help\n"
+		    "  -w <num>  Wrap lines to have at most <num> bytes\n"
+		    "  -d <str>  Specify punctuation to break on. Order determines preference.\n"
+		    "  -s        Skip passing punctuation around wrapping points to the command\n";
+	return 1;
+}
+
+vector<UChar32> parse_delimiters(char *value) {
+	int32_t length = strlen(value);
+	int32_t pos = 0;
+	
+	vector<UChar32> delimiters;
+	delimiters.reserve(length);
+
+	while (pos < length) {
+		UChar32 delimiter;
+		U8_NEXT(value, pos, length, delimiter);
+		
+		if (delimiter < 0)
+			throw utf8::NotUTF8Exception(value);
+		
+		delimiters.push_back(delimiter);
+	}
+
+	return delimiters;
+}
+
+void parse_options(program_options &options, int argc, char **argv) {
+	while (true) {
+		switch(getopt(argc, argv, "+w:d:sh")) {
+			case 'w':
+				options.column_width = atoi(optarg);
+				continue;
+
+			case 'd':
+				options.delimiters = parse_delimiters(optarg);
+				continue;
+
+			case 's':
+				options.keep_delimiters = false;
+				continue;
+
+			case 'h':
+			case '?':
+			default:
+				exit(usage(argv));
+
+			case -1:
+				break;
+		}
+		break;
+	}
+
+	if (optind == argc)
+		exit(usage(argv));
+
+	options.child_argv = argv + optind;
+}
+
+int main(int argc, char **argv) {
+	program_options options;
+
+	parse_options(options, argc, argv);
+
+	util::UnboundedSingleQueue<deque<string>> queue;
```

Comment:
The queue wasn't really designed with big objects in mind.  It does a simple `operator=` copy in `Produce` and in `Consume`.  Consider some sort of `swap` or `emplace` semantics.

Reply:
I've added move semantics to UnboundedSingleQueue.